Rework linux launching scripts to reflect conda

peterfpeterson opened the pull request:
This really is a cleanup of the scripts. It drops support for qt4, removes mention of mantidplot, and allows for taking advantage of conda.

In conda mode: let conda take care of the PYTHONPATH, don't do it in the scripts. This also re-uses CONDA_PREFIX which points at the location of the active environment. This does verify that the conda package is being used from an active conda environment.

In system mode: do what we've always done.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

Run `cmake .` and compare the build version, install version, and previous install version of `mantidworkbench` and `mantidpython`.

*There is no associated issue.*


*This does not require release notes* because it should not affect users in any noticeable way.
<!-- delete this if you added release notes
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

martyngigg commented on buildconfig/CMake/LinuxPackageScripts.cmake:
```diff
@@ -131,6 +131,19 @@ endif()
 # ######################################################################################################################
+set(CONDA_PREAMBLE_TEXT
+    "# Verify that conda is setup
+if [ -z \"\${CONDA_PREFIX}\" ]; then
```
We've been using this script without an activation of the conda environment on our analysis machines. We currently have a launcher menu item pointed at it.
It also won't work in the tarred up bundle we're creating:  https://github.com/mantidproject/mantid/releases/download/nightly/mantidworkbenchnightly-6.4.20220907.1204.tar.xz as that doesn't have a full conda environment

Can we just say use the `CONDA_PREFIX` if defined else fallback to the relative path checks?

peterfpeterson replied:
I did not know anything about how that tarball works and assumed it acted like a proper conda package. I'll do some more investigation.

martyngigg commented on buildconfig/CMake/Packaging/launch_mantidworkbench.sh.in:
```diff
@@ -1,34 +1,23 @@
 #!/bin/sh
 #
-# Launch Mantidplot using any necessary LD_PRELOAD or software collection behaviour
```
There's a fair few MantidPlot references still here. While we're editing can we get rid of them?

peterfpeterson commented:
Real questions: Should the script be split into development is hooked onto `CONDA_ENV` and production connected to `CONDA_BUILD`? The code needs more work, but I think that is the distinction we need (assuming I understand the variables correctly).

peterfpeterson committed:
Remove check for qt4

peterfpeterson committed:
Conditionally set python executable to use

peterfpeterson committed:
Refactor how workbench is started

martyngigg commented on buildconfig/CMake/LinuxPackageScripts.cmake:
```diff
@@ -214,11 +239,20 @@ execute_process(COMMAND "chmod" "+x" "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/AddPytho
+  set(MANTIDWORKBENCH_EXEC "\${CONDA_PREFIX}/bin/workbench") # what the actual thing is called
+else()
+  set(LOCAL_PYPATH "\${INSTALLDIR}/bin:\${INSTALLDIR}/lib:\${INSTALLDIR}/plugins")
+  set(PYTHON_EXEC_LOCAL "${PYTHON_EXECUTABLE}")
```
Should this be `set(PYTHON_EXEC_LOCAL "\${PYTHON_EXECUTABLE}")` so that the `PYTHON_EXECUTABLE` is expanded later?

peterfpeterson replied:
`PYTHON_EXECUTABLE` is the value from cmake so it should be expanded here.

martyngigg commented on buildconfig/CMake/LinuxPackageScripts.cmake:
```diff
@@ -214,11 +239,20 @@ execute_process(COMMAND "chmod" "+x" "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/AddPytho
-# used by mantidplot and mantidworkbench
-set(LOCAL_PYPATH "\${INSTALLDIR}/bin:\${INSTALLDIR}/lib:\${INSTALLDIR}/plugins")
+# used by mantidworkbench
+if(CONDA_BUILD)
```
I just need to check how we are launching workbench on our analysis machines as I'm not sure they have `conda run` before them.

peterfpeterson replied:
We are activating the conda environment before running the `mantidworkbench` script.

It is moderately convoluted. I didn't know that `conda run` was a thing. Thanks for the tip!

martyngigg reviewed (changes requested):
I've just got one comment that I think a dollar needs escaping.

peterfpeterson commented:
Documenting conversation with @martyngigg. The new logic (for `CONDA_BUILD=on`) is

Get install location (current `INSTALL_DIR`)
  * if it matches `CONDA_PREFIX` then launch mantid as usual
  * else if `conda` exists `conda run -n <conda environment> mantidworkbench $@[*]`
  * else error out

peterfpeterson commented:
Variables that are useful for this:
* `CONDA_DEFAULT_ENV` points at the name of the conda environment that is active (e.g. `mantid-dev`)
* `CONDA_PREFIX` contains the full path of the current conda environment (e.g. `/opt/anaconda/envs/mantid-dev`)

[Related branch](https://github.com/mantidproject/mantid/compare/main...peterfpeterson:mantid:cleanup_launchers_v2)

martyngigg committed:
Create to workbench lauchers; 1 for conda, 1 for standalone

The standalone-based installation takes the latter where
it doesn't have a complete conda installation.

martyngigg commented:
The RHEL7 build is no longer relevant so it can be ignored when considering whether to gatekeeper this.
